server/wsconnection.go

```diff
@@ -35,5 +37,9 @@ func (c wsConnection) Send(_ context.Context, message *protobufs.ServerToAgent)
 }
 
 func (c wsConnection) Disconnect() error {
+	if c.closed.Load() {
+		return nil
+	}
+	c.closed.Store(true)
```

Comment:
To ensure this happens once, I think the correct way is to use the atomic `CompareAndSwap` instead of separate Load/Store which can race.

Reply:
Good call